File: metabase_pipeline/metabase_client.py
"""
metabase_client.py — Metabase API client
Supports session token auth (from browser) with username/password fallback.
"""

import requests
import json
import csv
import os
from typing import Optional
from config import BASE_URL, SESSION_TOKEN, METABASE_USER, METABASE_PASS, DEFAULT_QUERY_BODY


class MetabaseClient:
    def __init__(self):
        self.base_url = BASE_URL.rstrip("/")
        self.session = requests.Session()
        self._setup_headers(SESSION_TOKEN)

    def _setup_headers(self, token: str):
        self.session_token = token
        self.session.headers.update({
            "X-Metabase-Session": token,
            "Content-Type": "application/json",
            "Accept": "application/json",
        })

    # ─────────────────────────────────────────────
    # Authentication
    # ─────────────────────────────────────────────

    def authenticate(self) -> str:
        """Login with username/password and refresh session token."""
        if not METABASE_USER or not METABASE_PASS:
            raise ValueError("Set METABASE_USER and METABASE_PASS in config.py for re-authentication.")
        r = self.session.post(f"{self.base_url}/api/session", json={
            "username": METABASE_USER,
            "password": METABASE_PASS,
        })
        r.raise_for_status()
        token = r.json()["id"]
        self._setup_headers(token)
        print(f"✅ Authenticated via username/password.")
        return token

    def check_session(self) -> bool:
        """Check if the current session token is still valid."""
        r = self.session.get(f"{self.base_url}/api/user/current")
        if r.status_code == 200:
            user = r.json()
            print(f"✅ Session valid — logged in as: {user.get('email', 'unknown')}")
            return True
        else:
            print(f"⚠️  Session expired or invalid (status {r.status_code})")
            return False

    def ensure_authenticated(self):
        """Check session; re-authenticate with credentials if expired."""
        if not self.check_session():
            print("🔄 Attempting re-authentication...")
            self.authenticate()

    def logout(self):
        self.session.delete(f"{self.base_url}/api/session")
        print("👋 Logged out.")

    # ─────────────────────────────────────────────
    # Core Query Runner
    # ─────────────────────────────────────────────

    def run_card_json(self, card_id: int, parameters: list = None) -> list[dict]:
        """
        Execute a saved question and return results as list of row dicts.

        Args:
            card_id:    Metabase card/question ID
            parameters: Optional filter parameters list

        Returns:
            List of dicts, one per row
        """
        body = {**DEFAULT_QUERY_BODY, "parameters": parameters or []}
        url = f"{self.base_url}/api/card/{card_id}/query/json"

        r = self.session.post(url, json=body)

        if r.status_code == 401:
            print("⚠️  401 — session expired, re-authenticating...")
            self.authenticate()
            r = self.session.post(url, json=body)

        r.raise_for_status()
        data = r.json()
        print(f"✅ Card {card_id}: {len(data)} rows returned.")
        return data

    def run_card_csv(self, card_id: int, parameters: list = None) -> str:
        """Execute a saved question and return raw CSV string."""
        body = {**DEFAULT_QUERY_BODY, "parameters": json.dumps(parameters or [])}
        url = f"{self.base_url}/api/card/{card_id}/query/csv"

        r = self.session.post(url, data=body)
        if r.status_code == 401:
            self.authenticate()
            r = self.session.post(url, data=body)

        r.raise_for_status()
        return r.text

    # ─────────────────────────────────────────────
    # Discovery
    # ─────────────────────────────────────────────

    def list_questions(self) -> list[dict]:
        r = self.session.get(f"{self.base_url}/api/card")
        r.raise_for_status()
        return r.json()

    def list_dashboards(self) -> list[dict]:
        r = self.session.get(f"{self.base_url}/api/dashboard")
        r.raise_for_status()
        return r.json()

    def list_databases(self) -> list[dict]:
        r = self.session.get(f"{self.base_url}/api/database")
        r.raise_for_status()
        return r.json().get("data", [])

    def run_native_query(self, database_id: int, sql: str) -> dict:
        """Run a raw SQL query against a connected database."""
        r = self.session.post(f"{self.base_url}/api/dataset", json={
            "database": database_id,
            "type": "native",
            "native": {"query": sql},
        })
        r.raise_for_status()
        result = r.json()
        cols = [col["name"] for col in result["data"]["cols"]]
        rows = result["data"]["rows"]
        return {"columns": cols, "rows": [dict(zip(cols, row)) for row in rows]}

File: src/integrations/metabase/metabase.client.ts
/**
 * metabase.client.ts — Metabase API client
 * Supports session token auth (from browser) with username/password fallback.
 */

import { env } from '@config/env.validation';
import { logger } from '@common/utils/logger.util';

const BASE_URL = 'https://plmb.shopup.center';

const DEFAULT_QUERY_BODY = {
  parameters: [],
  format_rows: true,
  pivot_results: false,
};

export interface MetabaseRow {
  [column: string]: unknown;
}

export interface NativeQueryResult {
  columns: string[];
  rows: MetabaseRow[];
}

export class MetabaseClient {
  private readonly baseUrl: string;
  private sessionToken: string;

  constructor() {
    this.baseUrl = BASE_URL;
    this.sessionToken = env.METABASE_SESSION_TOKEN ?? '';
  }

  private headers(): Record<string, string> {
    return {
      'X-Metabase-Session': this.sessionToken,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };
  }

  private async get<T>(path: string): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'GET',
      headers: this.headers(),
    });
    if (!res.ok) throw new Error(`Metabase GET ${path} failed: ${res.status}`);
    return res.json() as Promise<T>;
  }

  private async post<T>(path: string, body: unknown, asForm = false): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: asForm
        ? { 'X-Metabase-Session': this.sessionToken }
        : this.headers(),
      body: asForm
        ? new URLSearchParams(body as Record<string, string>)
        : JSON.stringify(body),
    });
    if (!res.ok) throw new Error(`Metabase POST ${path} failed: ${res.status}`);
    return res.json() as Promise<T>;
  }

  // ─────────────────────────────────────────────
  // Authentication
  // ─────────────────────────────────────────────

  /** Login with username/password and refresh session token. */
  async authenticate(): Promise<string> {
    const user = env.METABASE_USER;
    const pass = env.METABASE_PASS;
    if (!user || !pass) {
      throw new Error('Set METABASE_USER and METABASE_PASS in .env for re-authentication.');
    }
    const data = await this.post<{ id: string }>('/api/session', {
      username: user,
      password: pass,
    });
    this.sessionToken = data.id;
    logger.info('Metabase: authenticated via username/password');
    return this.sessionToken;
  }

  /** Check if the current session token is still valid. */
  async checkSession(): Promise<boolean> {
    try {
      const user = await this.get<{ email?: string }>('/api/user/current');
      logger.info(`Metabase: session valid — logged in as ${user.email ?? 'unknown'}`);
      return true;
    } catch {
      logger.warn('Metabase: session expired or invalid');
      return false;
    }
  }

  /** Check session; re-authenticate with credentials if expired. */
  async ensureAuthenticated(): Promise<void> {
    const valid = await this.checkSession();
    if (!valid) {
      logger.info('Metabase: attempting re-authentication...');
      await this.authenticate();
    }
  }

  async logout(): Promise<void> {
    await fetch(`${this.baseUrl}/api/session`, {
      method: 'DELETE',
      headers: this.headers(),
    });
    logger.info('Metabase: logged out');
  }

  // ─────────────────────────────────────────────
  // Core Query Runner
  // ─────────────────────────────────────────────

  /**
   * Execute a saved question and return results as an array of row objects.
   * Retries once if the session has expired (401).
   */
  async runCardJson(cardId: number, parameters: unknown[] = []): Promise<MetabaseRow[]> {
    const body = { ...DEFAULT_QUERY_BODY, parameters };
    const path = `/api/card/${cardId}/query/json`;

    let res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: this.headers(),
      body: JSON.stringify(body),
    });

    if (res.status === 401) {
      logger.warn('Metabase: 401 on runCardJson — re-authenticating...');
      await this.authenticate();
      res = await fetch(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers: this.headers(),
        body: JSON.stringify(body),
      });
    }

    if (!res.ok) throw new Error(`Metabase card ${cardId} query failed: ${res.status}`);
    const data = (await res.json()) as MetabaseRow[];
    logger.info(`Metabase: card ${cardId} returned ${data.length} rows`);
    return data;
  }

  /** Execute a saved question and return raw CSV string. */
  async runCardCsv(cardId: number, parameters: unknown[] = []): Promise<string> {
    const path = `/api/card/${cardId}/query/csv`;
    const formBody = new URLSearchParams({
      parameters: JSON.stringify(parameters),
      format_rows: String(DEFAULT_QUERY_BODY.format_rows),
      pivot_results: String(DEFAULT_QUERY_BODY.pivot_results),
    });

    let res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: { 'X-Metabase-Session': this.sessionToken },
      body: formBody,
    });

    if (res.status === 401) {
      await this.authenticate();
      res = await fetch(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers: { 'X-Metabase-Session': this.sessionToken },
        body: formBody,
      });
    }

    if (!res.ok) throw new Error(`Metabase card ${cardId} CSV query failed: ${res.status}`);
    return res.text();
  }

  // ─────────────────────────────────────────────
  // Discovery
  // ─────────────────────────────────────────────

  listQuestions(): Promise<MetabaseRow[]> {
    return this.get<MetabaseRow[]>('/api/card');
  }

  listDashboards(): Promise<MetabaseRow[]> {
    return this.get<MetabaseRow[]>('/api/dashboard');
  }

  async listDatabases(): Promise<MetabaseRow[]> {
    const res = await this.get<{ data: MetabaseRow[] }>('/api/database');
    return res.data;
  }

  /** Run a raw SQL query against a connected database. */
  async runNativeQuery(databaseId: number, sql: string): Promise<NativeQueryResult> {
    const res = await this.post<{
      data: { cols: { name: string }[]; rows: unknown[][] };
    }>('/api/dataset', {
      database: databaseId,
      type: 'native',
      native: { query: sql },
    });

    const columns = res.data.cols.map((c) => c.name);
    const rows = res.data.rows.map((row) =>
      Object.fromEntries(columns.map((col, i) => [col, row[i]])),
    );
    return { columns, rows };
  }
}
